controller/snapshot_controller.go

```diff
@@ -0,0 +1,584 @@
+			snapshot.Status.Error = "lost track of the corresponding snapshot info inside volume engine"
+		}
+		// Newly created snapshotCR, wait for the snapshotInfo to be appeared inside engine.Status.Snapshot
+		snapshot.Status.ReadyToUse = false
+		return nil
+	}
+
+	if err := syncSnapshotWithSnapshotInfo(snapshot, snapshotInfo, engine.Spec.VolumeSize); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (sc *SnapshotController) generatingEventsForSnapshot(existingSnapshot, snapshot *longhorn.Snapshot) {
+	if !existingSnapshot.Status.MarkRemoved && snapshot.Status.MarkRemoved {
+		sc.eventRecorder.Event(snapshot, v1.EventTypeWarning, "SnapshotDelete", "snapshot is marked as removed")
+	}
+	if snapshot.Spec.CreateSnapshot && existingSnapshot.Status.CreationTime == "" && snapshot.Status.CreationTime != "" {
+		sc.eventRecorder.Eventf(snapshot, v1.EventTypeNormal, "SnapshotCreate", "successfully provisioned the snapshot")
+	}
+	if snapshot.Status.Error != "" && existingSnapshot.Status.Error != snapshot.Status.Error {
+		sc.eventRecorder.Eventf(snapshot, v1.EventTypeWarning, "SnapshotError", "%v", snapshot.Status.Error)
+	}
+	if existingSnapshot.Status.ReadyToUse != snapshot.Status.ReadyToUse {
+		if snapshot.Status.ReadyToUse {
+			sc.eventRecorder.Eventf(snapshot, v1.EventTypeNormal, "SnapshotUpdate", "snapshot becomes ready to use")
+		} else {
+			sc.eventRecorder.Eventf(snapshot, v1.EventTypeWarning, "SnapshotUpdate", "snapshot becomes not ready to use")
+		}
+	}
+}
+
+func syncSnapshotWithSnapshotInfo(snap *longhorn.Snapshot, snapInfo *longhorn.SnapshotInfo, restoreSize int64) error {
+	size, err := strconv.ParseInt(snapInfo.Size, 10, 64)
+	if err != nil {
+		return err
+	}
+	snap.Status.Parent = snapInfo.Parent
+	snap.Status.Children = snapInfo.Children
+	snap.Status.MarkRemoved = snapInfo.Removed
+	snap.Status.UserCreated = snapInfo.UserCreated
+	snap.Status.CreationTime = snapInfo.Created
+	snap.Status.Size = size
+	snap.Status.Labels = snapInfo.Labels
+	snap.Status.RestoreSize = restoreSize
+	if snap.Status.MarkRemoved || snap.DeletionTimestamp != nil {
+		snap.Status.ReadyToUse = false
+	} else {
+		snap.Status.ReadyToUse = true
+		snap.Status.Error = ""
+	}
+	return nil
+}
+
+func (sc *SnapshotController) getTheOnlyEngineCRforSnapshot(snapshot *longhorn.Snapshot) (*longhorn.Engine, error) {
+	engines, err := sc.ds.ListVolumeEngines(snapshot.Spec.Volume)
+	if err != nil {
+		return nil, errors.Wrap(err, "getTheOnlyEngineCRforSnapshot")
+	}
+	if len(engines) != 1 {
+		return nil, fmt.Errorf("getTheOnlyEngineCRforSnapshot: found more than 1 engines for volume %v", snapshot.Spec.Volume)
+	}
+	var engine *longhorn.Engine
+	for _, e := range engines {
+		engine = e
+		break
+	}
+	return engine, nil
+}
+
+func (sc *SnapshotController) handleSnapshotCreate(snapshot *longhorn.Snapshot, engine *longhorn.Engine) error {
+	client, err := GetClientForEngine(engine, sc.engineClientCollection, engine.Status.CurrentImage)
```

Comment:
This is probably needed to change to use proxy client instead after IM proxy PR merged. The same for others related to the communication with engine process. 

cc Chinya and Shuo

Reply:
I changed it to use the proxy client. Please take a look @c3y1huang